syntax = "proto3";

package trace; // TODO need to change package name, it's confusing w/ class name

enum OrderingType {
	TIME = 0;
	DEPENDENCY = 1;
	WEIGHT = 2;
	NONE = 3;
}

enum EntryType {
	RECEIVE = 0;
	ASSIGN = 1;
	ACCEPT = 2;
	MOOT = 3; // TODO Fix the weird naming scheme
}

message StartTrace { // TODO: Change name to TraceHeader
	uint64 node_id = 1;
	OrderingType type = 2;
	uint64 start_time = 3;
}

message TraceRecord {
	uint64 trace_time = 1;
	EntryType type = 2;
	uint64 leader_id = 3;
	string msg = 4;
	uint64 slot = 5;
	uint64 client_id = 6;
	optional uint64 thread_id = 7; //Uniquely identifying the execution context
}

message MessageInfo {
	string msg = 1; // TODO: Change this eventually! Use templates?
	uint64 slot = 2; // TODO: Change this! Abstract away ordering?
	uint64 client_id = 3; // Client who sent the message
}
